src/ucp/core/ucp_mm.c

```diff
@@ -686,7 +687,107 @@ void ucp_mpool_obj_init(ucs_mpool_t *mp, void *obj, void *chunk)
 {
     ucp_mem_desc_t *elem_hdr  = obj;
     ucp_mem_desc_t *chunk_hdr = (ucp_mem_desc_t*)((ucp_mem_desc_t*)chunk - 1);
-    elem_hdr->memh = chunk_hdr->memh;
+    ucp_rndv_mpool_priv_t *mpriv;
+
+    mpriv              = (ucp_rndv_mpool_priv_t *)ucs_mpool_priv(mp);
+    elem_hdr->memh     = chunk_hdr->memh;
+    elem_hdr->ptr      = (void*)((ucp_mem_desc_t *)obj + 1);
+    elem_hdr->mem_type = mpriv->mem_type;
+}
+
+static inline ucs_status_t
+ucp_rndv_frag_malloc_mpools(ucs_mpool_t *mp, size_t *size_p, void **chunk_p)
+{
+    ucp_worker_h worker;
+    ucp_mem_desc_t *chunk_hdr;
+    ucp_mem_h memh;
+    ucp_mem_h meta_memh;
+    ucs_status_t status;
+    ucp_mem_map_params_t mem_params;
+    ucp_rndv_mpool_priv_t *mpriv;
+    int num_frags, frag_size;
+
+    /* Need to get default flags from ucp_mem_map_params2uct_flags() */
+    mem_params.field_mask = 0;
+
+    mpriv                 = (ucp_rndv_mpool_priv_t *)ucs_mpool_priv(mp);
+    worker                = mpriv->worker;
+    num_frags             = mpriv->num_frags;
+    frag_size             = mpriv->frag_size;
+
+    /* metadata */
+    status = ucp_mem_map_common(worker->context, NULL,
+                                (*size_p + sizeof(*chunk_hdr)),
+                                UCS_MEMORY_TYPE_HOST,
+                                ucp_mem_map_params2uct_flags(&mem_params),
+                                1, ucs_mpool_name(mp), &meta_memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    /* payload */
+    status = ucp_mem_map_common(worker->context, NULL,
+                                (frag_size * num_frags), mpriv->mem_type,
+                                ucp_mem_map_params2uct_flags(&mem_params),
+                                1, ucs_mpool_name(mp), &memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    chunk_hdr            = meta_memh->address;
+    chunk_hdr->meta_memh = meta_memh;
+    chunk_hdr->memh      = memh;
+    chunk_hdr->count     = 0;
```

Comment:
why needed?
you could calc the index by pointer arithmetic : (metadata - chunk_hdr) / (metadata_size + mpool obj size)
though, make sure mpool is creating with align==1